agent_test.go

```diff
@@ -97,9 +97,12 @@ func TestAgent_registerServiceAndCheck(t *testing.T) {
 	retry.Run(t, ensureRegistered)
 
 	// Deregister the service
-	if err := agent.client.Agent().ServiceDeregister(serviceID); err != nil {
-		t.Fatal(err)
-	}
+	go func() {
+		time.Sleep(time.Second)
+		if err := agent.client.Agent().ServiceDeregister(serviceID); err != nil {
+			t.Fatal(err)
+		}
+	}()
```

Comment:
It looks you are adding concurrency and paused for one second. I'm curious about the reason for this. Intuitively, it seems like it would be better to call Deregister immediately (rather than wait a second) and since we want to wait for this call to finish, it seems like extra work to make it concurrent. Can the tests run successfully without this or are they maybe part of tuning the `retry.RunWith`?

Reply:
We want to start checking for the service deregistration before the actual deregistration() is called, that's why there's the Sleep call there. That gives us time (hopefully) to start `ensureDeregistered`.

Previously it could happen that the service was deregistered, and registered again by ESM, even before we started waiting for the registration to be gone. Which would cause is to wait forever and fail.